lib/view_component/collection.rb

```diff
@@ -0,0 +1,54 @@
+
+# frozen_string_literal: true
+
+module ViewComponent
+  class Collection
+    def initialize(component, object = nil, options)
+      @component = component
+      @object = object
+      @options = options
+    end
+
+    def render_in(view_context, &block)
+      as = as_variable(@component.virtual_path, @options)
+      collection = collection_variable(@object, @options)
+      args = @options.except(:collection, :as)
+
+      collection.map do |item|
+        args[as] = item
+        @component.new(args).render_in(view_context, &block)
+      end.join
+    end
+
+    private
+
+    def collection_variable(object, options)
+      if object.respond_to?(:to_ary)
+        object.to_ary
+      elsif options.key?(:collection)
+        collection = options[:collection]
+        collection ? collection.to_a : []
+      else
+        raise ArgumentError.new("Must specify the option `collection` or pass a valid collection object.")
+      end
+    end
+
+    # Copied from https://github.com/rails/rails/blob/e2cf0b1d780b2e09f5270249ca021d94ce4fff9d/actionview/lib/action_view/renderer/partial_renderer.rb
+    def as_variable(path, options)
+      if as = options[:as]
+        raise_invalid_option_as(as) unless /\A[a-z_]\w*\z/.match?(as.to_s)
+        as.to_sym
+      else
+        File.basename(path).gsub(/_component/, "").to_sym
```

Comment:
I don't think you need to use the path to do this. Unlike ActionView partials we have a concrete class that we can leverage to determine the default variable. You could do 

```ruby
class.name.demodulize.underscore.chomp("_component").to_sym
```

Reply:
Do we need this variable to be dynamic?  Unlike ActionView partials we have concrete classes that we can hand this kind of behavior off.  Wouldn't it be simple to define this in the component using a class_attribute? Something like this:


```ruby
class ProductCouponComponent < ViewComponent::Base

  with_collection_parameter :item # if you don't call this would default to :product_coupon

  def initialize(item:)
    @item = item
  end
end
```

Reply:
I really like this idea, especially because it mean removing `as:` as one of the special values in the options hash. Implemented in the latest on this branch (along with directly using the demodulized class name of the component).